File: topic3/code/code/test_tworings.m
clear; 

%%
% generate the data; 

% generate random angles between [0, 2pi]
n = 800; 
rangle = 2 * pi * rand(n, 1); 

% generate random radius for the first circle; 
e = 0.2;
rr = 1.9 + e * rand(n, 1); 

rx = rr .* sin(rangle); 
ry = rr .* cos(rangle); 

x = rx; 
y = ry; 

% generate random radius for the second circle; 
rr2 = 1.2 + e * rand(n, 1); 

rx2 = rr2 .* sin(rangle); 
ry2 = rr2 .* cos(rangle); 

x = [x; rx2]; 
y = [y; ry2]; 

rx3 = 1.4 + (1.9 - 1.4) * rand(10, 1); 
ry3 = e * rand(10, 1); 

% % uncomment this to comment the two rings; 
x = [x; rx3]; 
y = [y; ry3]; 

%%
data = [x, y]; 

figure; 
plot(x, y, 'k.'); 
hold on; 
title('original data'); 

keyboard; 

% run kmeans on the original coordinates; 
K = 2; 
idx = kmeans([x, y], K, 'Replicates', 10); 

figure; 
plot(x(idx==1), y(idx==1), 'r.'); 
hold on; 
plot(x(idx==2), y(idx==2), 'b.'); 
title('K-means'); 

keyboard;

%%
distmat = squareform(pdist(data)).^2; 

% A(A<0.99) = 0; 
A = double(distmat < 0.1);  

% figure; 
% spy(A); 
% hold on; 
% title('adjacency matrix'); 

%%

D = diag(sum(A,2)); 
L = D - A; 

[V, S] = eig(L); 
K = 2; 
idx = kmeans(V(:,1:2), K, 'Replicates', 10); 

figure; 
plot(x(idx==1), y(idx==1), 'r.'); 
hold on; 
plot(x(idx==2), y(idx==2), 'b.'); 
title('Spectral Clustering'); 

File: topic3/code/python/test_tworings.py
import numpy as np
from matplotlib import pyplot as plt
from sklearn.cluster import KMeans
from sklearn.metrics import pairwise_distances

# generate the data
# generate random angles between [0, 2pi]
n = 800
rangle = 2 * np.pi * np.random.rand(n, 1)

# generate random radius for the first circle
e = 0.2
rr = 1.9 + e * np.random.rand(n, 1)

rx = rr * np.sin(rangle)
ry = rr * np.cos(rangle)

x = rx
y = ry

# generate random radius for the second circle
rr2 = 1.2 + e * np.random.rand(n, 1)

rx2 = rr2 * np.sin(rangle)
ry2 = rr2 * np.cos(rangle)

x = np.concatenate((x, rx2))
y = np.concatenate((y, ry2))

rx3 = 1.4 + (1.9 - 1.4) * np.random.rand(10, 1)
ry3 = e * np.random.rand(10, 1)

# uncomment this to comment the two rings;
x = np.concatenate((x, rx3))
y = np.concatenate((y, ry3))

data = np.concatenate((x, y), axis=1)

plt.scatter(data[:, 0], data[:, 1], c='black')
plt.title('original data')
plt.show()

# run kmeans on the original coordinates
K = 2
kmeans = KMeans(n_clusters=K).fit(data)
idx = kmeans.labels_

plt.scatter([x for i, x in enumerate(data[:, 0]) if idx[i] == 0],
            [y for i, y in enumerate(data[:, 1]) if idx[i] == 0],
            c='r')

plt.scatter([x for i, x in enumerate(data[:, 0]) if idx[i] == 1],
            [y for i, y in enumerate(data[:, 1]) if idx[i] == 1],
            c='b')

plt.title('K-means')
plt.show()

distmat = pairwise_distances(data) * pairwise_distances(data)

A = (distmat < 0.1).astype(np.int)

plt.spy(A)
plt.title('Adjacency Matrix')
plt.show()

D = np.diag(np.sum(A, axis=1))
L = D - A

s, v = np.linalg.eig(L)
K = 2
v = v[:, 0:K].real
kmeans = KMeans(n_clusters=K).fit(v)
idx = kmeans.labels_

plt.scatter([x for i, x in enumerate(data[:, 0]) if idx[i] == 0],
            [y for i, y in enumerate(data[:, 1]) if idx[i] == 0],
            c='r')

plt.scatter([x for i, x in enumerate(data[:, 0]) if idx[i] == 1],
            [y for i, y in enumerate(data[:, 1]) if idx[i] == 1],
            c='b')

plt.title('Spectral Clustering')
plt.show()
